include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,71 +68,159 @@ module BYTES-SIMPLIFICATION [symbolic]
 
     // #range
 
-    rule #range(_, _, K) => .Bytes requires K <=Int 0 [simplification]
+    rule [range-empty]:
+      #range(_:Bytes, S:Int, W:Int) => .Bytes
+      requires S <Int 0 orBool W <=Int 0
+      [simplification(30)]
 
-    rule #range( M [ N := BUF:Bytes ], L, K ) =>
-         #let W = minInt(K, N -Int L) #in #range(M, L, W) +Bytes #range(M [ N := BUF ], N, K -Int W)
-      requires 0 <Int K andBool L <Int N [simplification]
+    rule [range-full]:
+      #range(B:Bytes, 0, W:Int) => B +Bytes #buf(W -Int lengthBytes(B), 0)
+      requires lengthBytes(B) <=Int W
+      [simplification]
+
+    rule [range-params-equal-ml]:
+      { #range (B1:Bytes, S1:Int, W1:Int) #Equals #range (B2:Bytes, S2:Int, W2:Int) } => #Top
+      requires B1 ==K B2 andBool S1 ==Int S2 andBool W1 ==Int W2
+      [simplification]
+
+    rule [range-included-in-cHead]:
+      #range(B1:Bytes +Bytes _:Bytes, S:Int, W:Int) => #range(B1, S, W)
+      requires S +Int W <=Int lengthBytes(B1)
+      [simplification]
+
+    rule [range-outside-cHead]:
+      #range(B1:Bytes +Bytes B2:Bytes, S:Int, W:Int) => #range(B2, S -Int lengthBytes(B1), W)
+      requires lengthBytes(B1) <=Int S
+      [simplification]
+
+    rule [range-includes-cHead]:
+      #range(B1:Bytes +Bytes B2:Bytes, 0, W:Int) => B1 +Bytes #range(B2, 0, W -Int lengthBytes(B1))
+      requires lengthBytes(B1) <=Int W
+      [simplification]
+
+    rule [range-inside-cHead-base-conc]:
+      #range(B:Bytes, S1:Int +Int S2:Int, W:Int) => #range(#range(B, S2, lengthBytes(B) -Int S2), S1, W)
+      requires 0 <=Int S1 andBool 0 <=Int S2 andBool 0 <Int S1 +Int S2 andBool S1 +Int S2 <=Int lengthBytes(B)
+      [concrete(B), concrete(S2), simplification]
```

Comment:
```suggestion
    rule [range-inside-cHead-base-conc]:
      #range(B:Bytes, S1:Int +Int S2:Int, W:Int) => #range(#range(B, S2, lengthBytes(B) -Int S2), S1, W)
      requires 0 <=Int S1 andBool 0 <=Int S2 andBool 0 <Int S1 +Int S2 andBool S1 +Int S2 <=Int lengthBytes(B)
      [concrete(B, S2), simplification]
```

Reply:
Ok, this is bizarre. When rewritten as you suggest, the claim
```
    claim [range-36]:
      <k> runLemma ( #range(B:Bytes, 5 +Int S2:Int, W:Int) ) => doneLemma ( #range(#range(B, 5, lengthBytes(B) -Int 5), S2, W) ) ... </k>
      requires 0 <=Int S2 andBool (5 +Int S2 <=Int lengthBytes(B))
```
does not pass because `B` is not concrete and the simplification is not triggered, and it shouldn't pass, actually. But originally, with the two `concrete`s separated, the simplification is somehow triggered. I'm not sure who should look at this, perhaps @radumereuta if frontend or @jberthold if backend?

Reply:
If you duplicate attributes on a rule, only one of them is taken into consideration.
```
// test.k
  rule A + B => A +Int B [concrete(A), concrete(B)]
// parsed.txt
  rule `_+__TEST-SYNTAX_Exp_Exp_Exp`(#SemanticCastToInt(A),#SemanticCastToInt(B))=>`_+Int_`(#SemanticCastToInt(A),#SemanticCastToInt(B)) requires #token("true","Bool") ensures #token("true","Bool")
   [concrete(B),
    org.kframework.attributes.Location(Location(10,8,10,25)),
    org.kframework.attributes.Source(Source(/home/radu/work/test/test.k)),
    org.kframework.definition.Production(syntax #RuleContent ::= #RuleBody [klabel(#ruleNoConditions),
    symbol])]
```
The front end should throw an error in this case. I'll make an issue.

Reply:
I've added a check in the front end: https://github.com/runtimeverification/k/pull/3438
You should get an error message from now on.